crypto/src/lib.rs

```diff
@@ -439,6 +439,27 @@ impl From<Multihash> for PublicKey {
     }
 }
 
+impl From<PrivateKey> for PublicKey {
+    #[allow(clippy::expect_used)]
+    fn from(private_key: PrivateKey) -> Self {
+        let digest_function = private_key.digest_function();
+        let key_gen_option = Some(UrsaKeyGenOption::FromSecretKey(UrsaPrivateKey(
+            private_key.payload,
+        )));
+        let (mut public_key, _) = match digest_function {
+            Algorithm::Ed25519 => Ed25519Sha512.keypair(key_gen_option),
+            Algorithm::Secp256k1 => EcdsaSecp256k1Sha256::new().keypair(key_gen_option),
+            Algorithm::BlsNormal => BlsNormal::new().keypair(key_gen_option),
+            Algorithm::BlsSmall => BlsSmall::new().keypair(key_gen_option),
+        }
+        .expect("can't fail for valid `PrivateKey`");
```

Comment:
We should actually fix this upstream in hyperledger/ursa

Reply:
If you mean filing a PR with this impl into their repo, then I'm a bit skeptical about the speed of their PR acceptance process, but I could try to.

Reply:
On further inspection, ursa has nothing but opaque bytearrays in place of private / public keys with no associated digests, and changing the structs to a product type would break FFI, macros, and who knows what else. So I guess the cost is not worth it.